Add Ceph RBD driver

codenrhoden opened the pull request:
This PR introduces the "rbd" storage driver, for interace with Ceph RBD devices.

This implementation does not use `go-ceph`, and therefore does not introduce new external dependencies. Instead it execs calls to the `rbd` executable, and parses the `json` output.

codenrhoden commented:
So far I've been rebasing this PR and and keeping the whole thing in one commit. That makes it hard to see that I'm adding new things, though. So far now, I'm going to push new commits with major new additions so you can see where I'm at. 

I think that makes piecemeal reviews easier as well.

We can squash everything together at the end.

My latest commit adds tests in the `tests` folder.  The tests are in line with all the other drivers (with some simplification and enhancements for the "inspect" tests).  I think only two missing pieces now are docs, and a greater, scripted functional test that involves more than one server.

Code review is welcome!

akutz commented:
Hi @codenrhoden,

I don't rebase during the review process because the GitHub code review doesn't work if you make in-line comments and then update the branch with a rebased copy. I always perform additional commits during review -- at least until the very end.

Sounds good Travis!

akutz commented on .docs/user-guide/config.md:
```diff
@@ -586,8 +586,9 @@ remote storage systems. Currently the following storage drivers are supported:
 [VirtualBox](./storage-providers.md#virtualbox) | virtualbox
-[EBS](./storage-providers.md#ebs) | ebs, ec2
-[EFS](./storage-providers.md#efs) | efs
+[EBS](./storage-providers.md#aws-ebs) | ebs, ec2
```
Good change, but please don't make it here as part of this PR. Please do it in a separate doc change.

akutz commented on .docs/user-guide/config.md:
```diff
@@ -586,8 +586,9 @@ remote storage systems. Currently the following storage drivers are supported:
-[EBS](./storage-providers.md#ebs) | ebs, ec2
-[EFS](./storage-providers.md#efs) | efs
+[EBS](./storage-providers.md#aws-ebs) | ebs, ec2
+[EFS](./storage-providers.md#aws-efs) | efs
```
Good change, but please don't make it here as part of this PR. Please do it in a separate doc change.

akutz commented on drivers/storage/rbd/tests/rbd_test.go:
```diff
@@ -0,0 +1,370 @@
+	return fmt.Sprintf("%s/24", a.ip.String())
+}
+
+var testIPs = []testIPInput{
```
Possibly optimize tests by parsing the unique IPs once and reusing the result.

akutz commented on drivers/storage/rbd/utils/utils.go:
```diff
@@ -0,0 +1,287 @@
+// +build !libstorage_storage_driver libstorage_storage_driver_rbd
```
Hi @codenrhoden,

I might add a sub-package called `rbd-utils` and make the file `rbd-utils.go` use the package main, thereby creating a new CLI. Then basically create a wrapper for this package so with the first arg of the CLI you can invoke a method from this file and the remaining args are the args for that function. 

I was just thinking it would be a quick and easy way to easily validate the functionality of the `utils` package.

codenrhoden commented:
I've removed the `WIP` designation on this PR, as I consider the work work to be feature-complete. I'm sure more things will be added, but it's ready for real review now. My latest commits added a `test-rbd` target to the `Makefile` for building the test binary, and also added a Vagrantfile and test plan in the `tests` directory.

codenrhoden commented:
@akutz.  I think we are all green here.  Take another look through when you can.

I did update the tests to use statically parsed IPs.

I can run my compiled tests package on the vagrant VM included in this PR as follows:

```
[vagrant@libstorage-rbd-test-admin ~]$ sudo ./rbd.test
PASS
coverage: 55.8% of statements in github.com/codedellemc/libstorage/drivers/storage/rbd, github.com/codedellemc/libstorage/drivers/storage/rbd/executor
```

codenrhoden committed:
Add Ceph RBD driver

This patch introduces the Ceph RBD driver
